### ISSUE TITLE
[P4-5] NLP/text pipeline template for text-based competitions

### ISSUE THREAD

**Issue by mikeycdavis** [mikeycdavis/bitgrit#41]
## Overview
Phase 4 currently has templates for regression (`exp(-RMSE)` metric, #25) and F1-optimized classification (imbalanced data, #26), but has no template for competitions involving natural language processing (NLP) or text-based features. Multiple past and potential future BitGrit competitions involve text data:

| Competition | Text Source | Task |
|---|---|---|
| AI Generated Text Classification | Article text | Binary: human vs AI-generated |
| Social Media Ethereum Price Prediction | Tweet text | Price direction from social signals |
| X (Twitter) Virality Prediction | Tweet text + metadata | Virality probability |

Without an NLP template, each text-based competition requires building text preprocessing, tokenization, and feature extraction from scratch.

**Depends on:** #20 (base class)
**Milestone:** Phase 4: Future Competition Readiness

### `templates/nlp_pipeline.py`

```python
class NLPPipeline(BitGritPipeline):
    """
    Template for BitGrit competitions involving text features.
    Supports TF-IDF, sentence embeddings, and hybrid (text + tabular) feature spaces.

    Usage:
        pipeline = NLPPipeline(
            competition_dir="ai_text_detection",
            target_col="label",
            id_col="id",
            text_col="text",          # REQUIRED: column containing the text to process
            n_classes=2,              # REQUIRED: number of target classes
            feature_strategy="tfidf"  # "tfidf", "embeddings", or "hybrid"
        )
    """
    metric = "f1"  # Most NLP competitions on BitGrit use F1
    require_string_submission = False

    def __init__(
        self,
        competition_dir: str,
        target_col: str,
        id_col: str,
        text_col: str,     # REQUIRED — column name containing text data
        n_classes: int,    # REQUIRED — number of target classes
        feature_strategy: str = "tfidf",  # "tfidf", "embeddings", or "hybrid"
    ):
        if feature_strategy not in ("tfidf", "embeddings", "hybrid"):
            raise ValueError(
                f"feature_strategy must be 'tfidf', 'embeddings', or 'hybrid', "
                f"got '{feature_strategy}'"
            )
        self.competition_dir = competition_dir
        self.target_col = target_col
        self.id_col = id_col
        self.text_col = text_col
        self.n_classes = n_classes
        self.feature_strategy = feature_strategy
```

#### 1. TF-IDF (`feature_strategy="tfidf"`)
- [ ] Text preprocessing function `preprocess_text(text: str) -> str`:
  - Lowercase
  - Remove URLs (regex: `https?://\S+`)
  - Remove HTML tags (regex: `<[^>]+>`)
  - Remove non-alphanumeric characters except spaces
  - Collapse multiple spaces to single space
  - Strip leading/trailing whitespace
- [ ] Feature extraction using `sklearn.feature_extraction.text.TfidfVectorizer`:
  ```python
  tfidf_params = {
      "max_features": 10000,      # cap vocabulary size
      "ngram_range": (1, 2),      # unigrams and bigrams
      "min_df": 3,                # ignore terms appearing in < 3 documents
      "max_df": 0.95,             # ignore terms appearing in > 95% of documents
      "sublinear_tf": True,       # apply log normalization to term frequencies
      "strip_accents": "unicode",
  }
  ```
- [ ] Fit TF-IDF on training text only. Transform both training and test text.
- [ ] Output: sparse matrix of shape `(n_samples, max_features)`

#### 2. Sentence embeddings (`feature_strategy="embeddings"`)
- [ ] Use `sklearn.feature_extraction.text.TfidfVectorizer` with `analyzer="word"` to create a weighted bag-of-words, then apply `sklearn.decomposition.TruncatedSVD(n_components=100)` to reduce dimensionality.
  - This is a lightweight embedding approach that does NOT require heavy dependencies like `sentence-transformers` or `torch`.
  - Rationale: BitGrit competitions typically have small datasets (< 50k rows) where heavy transformer models are overkill and may overfit.
- [ ] Pipeline: TF-IDF (max_features=50000) -> TruncatedSVD (n_components=100) -> dense numpy array
- [ ] Output: dense matrix of shape `(n_samples, 100)`

#### 3. Hybrid (`feature_strategy="hybrid"`)
- [ ] Combine TF-IDF/SVD text features with tabular features from non-text columns:
  ```python
  text_features = tfidf_svd_transform(df[text_col])  # (n, 100)
  tabular_features = df.drop(columns=[text_col, target_col, id_col]).values  # (n, k)
  combined = np.hstack([text_features, tabular_features])  # (n, 100 + k)
  ```
- [ ] Tabular features are passed through as-is (numeric columns assumed preprocessed by the competition-specific subclass)
- [ ] Text features are always SVD-reduced to 100 dimensions to prevent the text features from dominating the tabular features in the combined space

### Additional text-specific features (extracted for all strategies)
- [ ] `text_length`: character count of preprocessed text
- [ ] `word_count`: word count of preprocessed text
- [ ] `avg_word_length`: mean word length
- [ ] `unique_word_ratio`: number of unique words / total words
- [ ] `uppercase_ratio`: fraction of uppercase characters in original (pre-preprocessing) text
- [ ] `punctuation_count`: count of punctuation characters in original text
- [ ] `has_url`: binary — did the original text contain a URL (before removal)
- [ ] These 7 features are appended as additional columns to ALL feature strategies (TF-IDF, embeddings, hybrid)

### Ablation configs
```python
ABLATION_CONFIGS = [
    {
        "slot_name": "nlp_v1_tfidf",
        "model_types": ["lgbm", "xgb"],
        "feature_strategy": "tfidf",
        "ensemble_weights": "equal",
        "imbalance_strategy": "none",
    },
    {
        "slot_name": "nlp_v2_svd",
        "model_types": ["lgbm", "xgb"],
        "feature_strategy": "embeddings",
        "ensemble_weights": "equal",
        "imbalance_strategy": "none",
    },
    {
        "slot_name": "nlp_v3_hybrid",
        "model_types": ["lgbm", "xgb", "catboost"],
        "feature_strategy": "hybrid",
        "ensemble_weights": "equal",
        "imbalance_strategy": "none",
    },
    {
        "slot_name": "nlp_v4_weighted_hybrid",
        "model_types": ["lgbm", "xgb", "catboost"],
        "feature_strategy": "hybrid",
        "ensemble_weights": "oof_f1",
        "imbalance_strategy": "class_weight",
    },
]
```

### Why no transformer/deep learning models?
- BitGrit competition datasets are typically small (< 50k rows). TF-IDF + gradient boosting often outperforms fine-tuned transformers on small text datasets.
- No `torch`, `transformers`, or `sentence-transformers` dependency — keeps `requirements.txt` lightweight.
- If a future competition has large enough data to benefit from transformers, create a competition-specific override, not a template change.

### Dependencies
- [ ] No new dependencies required. All components use `scikit-learn` (already in requirements.txt):
  - `TfidfVectorizer` from `sklearn.feature_extraction.text`
  - `TruncatedSVD` from `sklearn.decomposition`
  - `Pipeline` from `sklearn.pipeline` (optional — for clean chaining)
- [ ] CatBoost note: CatBoost does not handle sparse matrices well. For TF-IDF strategy, convert to dense before passing to CatBoost, or exclude CatBoost from TF-IDF ablation slots (as shown in configs above — `nlp_v1_tfidf` uses only `lgbm` and `xgb`).

### Unit tests (`tests/test_nlp_template.py`)
- [ ] `test_text_col_required`: instantiating without `text_col` raises `TypeError`
- [ ] `test_n_classes_required`: instantiating without `n_classes` raises `TypeError`
- [ ] `test_invalid_feature_strategy_raises`: `feature_strategy="invalid"` raises `ValueError`
- [ ] `test_preprocess_text_removes_urls`: `preprocess_text("check https://example.com out")` returns `"check out"`
- [ ] `test_preprocess_text_removes_html`: `preprocess_text("<b>bold</b> text")` returns `"bold text"`
- [ ] `test_preprocess_text_lowercases`: `preprocess_text("HELLO World")` returns `"hello world"`
- [ ] `test_tfidf_features_shape`: TF-IDF on 50 samples produces shape `(50, min(max_features, vocab_size))`
- [ ] `test_svd_features_shape`: SVD reduction produces shape `(n_samples, 100)`
- [ ] `test_hybrid_features_combine_text_and_tabular`: hybrid features have `100 + n_tabular_cols + 7` columns (100 SVD + tabular + 7 text stats)
- [ ] `test_text_stat_features_computed`: `text_length`, `word_count`, etc. are present and numeric
- [ ] `test_tfidf_fit_on_train_only`: TF-IDF vocabulary is derived from training data only (test data may have OOV tokens that are ignored, not errored)
- [ ] `test_ablation_generates_4_csvs`: running on synthetic text data produces 4 CSV files

## Acceptance Criteria
- `NLPPipeline(competition_dir="test", target_col="y", id_col="id", text_col="text", n_classes=2)` instantiates without error
- Omitting `text_col` or `n_classes` raises `TypeError`
- Invalid `feature_strategy` raises `ValueError`
- TF-IDF is fit on training text only (never on test text)
- SVD produces exactly 100-dimensional dense features
- Hybrid mode combines 100 SVD features + tabular features + 7 text stat features
- No `torch`, `transformers`, or `sentence-transformers` dependency
- 4 ablation CSVs generated on synthetic text data
- All unit tests pass
- mypy passes on `templates/nlp_pipeline.py`
